src/tactic/rewrite_search/explain.lean

```diff
@@ -0,0 +1,255 @@
+  end
+
+/-- Explain a list of rewrites using `conv_x` tactics. -/
+meta def explain_rewrites (rs : list (expr × bool)) (s : side) (hows : list how) :
+  tactic string :=
+string.intercalate ",\n" <$> explanation_lines rs s none hows
+
+end using_conv
+
+private meta def explain_rewrites_concisely (steps : list (expr × bool)) (needs_refl : bool) :
+  tactic string :=
+do rules ← string.intercalate ", " <$> steps.mmap pp_rule,
+  return $ "erw [" ++ rules ++ "]" ++ (if needs_refl then ", refl" else "")
+
+/--
+Fails if we can't just use rewrite.
+Otherwise, returns 'tt' if we need a `refl` at the end.
+-/
+private meta def check_if_simple_rewrite_succeeds (rewrites : list (expr × bool)) (goal : expr) :
+  tactic bool :=
+lock_tactic_state $ do
+  m ← mk_meta_var goal,
+  set_goals [m],
+  rewrites.mmap' $ λ q, rewrite_target q.1 {symm := q.2, md := semireducible},
+  (reflexivity reducible >> return ff) <|> (reflexivity >> return tt)
+
+/-- Construct a list of rewrites from a proof unit. -/
+meta def proof_unit.rewrites (u : proof_unit) (rs : list (expr × bool)) : list (expr × bool) :=
+u.steps.filter_map $ how.to_rewrite rs
+
+/-- Construct an explanation string from a proof unit. -/
+meta def proof_unit.explain (u : proof_unit) (rs : list (expr × bool))
+  (explain_using_conv : bool) : tactic string :=
+if explain_using_conv then
+  using_conv.explain_rewrites rs u.side u.steps
+else
+  using_location.explain_rewrites rs u.side u.steps
+
+private meta def explain_proof_full (rs : list (expr × bool)) (explain_using_conv : bool) :
+list proof_unit → tactic string
+| [] := return ""
+| (u :: rest) := do
+  -- Don't use transitivity for the last unit, since it must be redundant.
+  head ← if rest.length = 0 ∨ u.side = side.L then pure [] else (do
+    n ← infer_type u.proof >>= (λ e, prod.snd <$> split_equation e) >>= pp,
+    pure $ ["transitivity " ++ to_string n]),
+  unit_expl ← u.explain rs explain_using_conv,
+  rest_expl ← explain_proof_full rest,
+  let expls := (head ++ [unit_expl, rest_expl]).filter $ λ t, ¬(t.length = 0),
+  return $ string.intercalate ",\n" expls
+
+private meta def explain_proof_concisely (rules : list (expr × bool)) (proof : expr)
+  (l : list proof_unit) : tactic string :=
+do let rws : list (expr × bool) := list.join $ l.map (λ u, do
+    (r, s) ← u.rewrites rules,
+    return (r, if u.side = side.L then s else ¬s)),
+  goal ← infer_type proof,
+  needs_refl ← check_if_simple_rewrite_succeeds rws goal,
+  explain_rewrites_concisely rws needs_refl
+
+/--
+Generate a human-readable explanation in Lean code of a proof generated by rewrite search.
+-/
+meta def explain_search_result (cfg : config) (rules : list (expr × bool)) (proof : expr)
+  (units : list proof_unit) : tactic string :=
+do explanation ← explain_proof_concisely rules proof units <|>
+                 explain_proof_full rules cfg.explain_using_conv units,
+  if cfg.explain then trace $ "/- `rewrite_search` says -/\n" ++ explanation
+  else skip,
```

Comment:
```suggestion
  when cfg.explain $ trace $ "/- `rewrite_search` says -/\n" ++ explanation,
```
Also, this should certainly be hooked up to the "Try this:" machinery. (See e.g. `squeeze_simp`)

Reply:
OK. TIL this existed, I've been a poor impoverished emacs user